File: python/test/test_basic/test_circuit_breaker.py
import asyncio
import threading

import pytest

from mental1104.concurrency.circuit_breaker import (
    CircuitBreaker,
    CircuitBreakerConfig,
    CircuitOpenError,
    CircuitOutcome,
    CircuitState,
    StateChangeReason,
)


class ManualClock:
    def __init__(self) -> None:
        self.value = 0.0
        self.lock = threading.Lock()

    def __call__(self) -> float:
        with self.lock:
            return self.value

    def advance(self, seconds: float) -> None:
        with self.lock:
            self.value += seconds


def config(**overrides):
    values = dict(
        window_seconds=10.0,
        minimum_requests=3,
        failure_rate_threshold=0.5,
        slow_call_duration_seconds=1.0,
        slow_call_rate_threshold=0.5,
        open_duration_seconds=5.0,
        half_open_max_probes=3,
        half_open_successes_to_close=3,
    )
    values.update(overrides)
    return CircuitBreakerConfig(**values)


def open_breaker(breaker: CircuitBreaker) -> None:
    for _ in range(3):
        breaker.try_acquire().record_failure()
    assert breaker.state is CircuitState.OPEN


def test_config_rejects_invalid_values():
    invalid = [
        {"window_seconds": 0},
        {"minimum_requests": 0},
        {"failure_rate_threshold": -0.1},
        {"failure_rate_threshold": 1.1},
        {"slow_call_duration_seconds": 0},
        {"slow_call_rate_threshold": -0.1},
        {"slow_call_rate_threshold": 1.1},
        {"open_duration_seconds": 0},
        {"half_open_max_probes": 0},
        {"half_open_successes_to_close": 0},
        {"half_open_max_probes": 2, "half_open_successes_to_close": 3},
    ]
    for overrides in invalid:
        with pytest.raises(ValueError):
            config(**overrides)


def test_minimum_requests_and_failure_rate_open_the_circuit():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)

    breaker.try_acquire().record_failure()
    breaker.try_acquire().record_success()
    assert breaker.state is CircuitState.CLOSED

    breaker.try_acquire().record_failure()
    assert breaker.state is CircuitState.OPEN


def test_slow_successes_can_open_the_circuit():
    clock = ManualClock()
    breaker = CircuitBreaker(
        config(
            minimum_requests=2,
            failure_rate_threshold=1.0,
            slow_call_rate_threshold=0.5,
        ),
        clock=clock,
    )

    breaker.try_acquire().record_success()
    permit = breaker.try_acquire()
    clock.advance(1.0)
    permit.record_success()

    snapshot = breaker.snapshot()
    assert snapshot.state is CircuitState.OPEN


def test_ignored_business_errors_do_not_enter_closed_window():
    clock = ManualClock()
    breaker = CircuitBreaker(config(minimum_requests=2), clock=clock)

    breaker.try_acquire().record_ignored()
    breaker.try_acquire().record_failure()

    snapshot = breaker.snapshot()
    assert snapshot.window_requests == 1
    assert snapshot.window_failures == 1
    assert snapshot.state is CircuitState.CLOSED


def test_sliding_window_prunes_expired_samples():
    clock = ManualClock()
    breaker = CircuitBreaker(
        config(window_seconds=2.0, minimum_requests=3),
        clock=clock,
    )

    breaker.try_acquire().record_failure()
    clock.advance(2.1)
    breaker.try_acquire().record_success()

    snapshot = breaker.snapshot()
    assert snapshot.window_requests == 1
    assert snapshot.window_failures == 0


def test_open_rejects_without_calling_operation_and_can_fallback():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)
    open_breaker(breaker)
    called = []

    def operation():
        called.append(True)
        return "downstream"

    with pytest.raises(CircuitOpenError):
        breaker.call(operation)
    assert not called

    value = breaker.call(operation, fallback=lambda _: "cached")
    assert value == "cached"
    assert not called


def test_half_open_allows_only_one_fixed_probe_round_then_closes():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)
    open_breaker(breaker)
    clock.advance(5.0)

    permits = [breaker.try_acquire() for _ in range(3)]
    assert breaker.state is CircuitState.HALF_OPEN
    with pytest.raises(CircuitOpenError):
        breaker.try_acquire()

    permits[0].record_success()
    permits[1].record_ignored()
    assert breaker.state is CircuitState.HALF_OPEN
    permits[2].record_success()

    assert breaker.state is CircuitState.CLOSED


def test_half_open_failure_or_slow_call_reopens():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)
    open_breaker(breaker)
    clock.advance(5.0)

    breaker.try_acquire().record_failure()
    assert breaker.state is CircuitState.OPEN

    clock.advance(5.0)
    permit = breaker.try_acquire()
    clock.advance(1.0)
    permit.record_success()
    assert breaker.state is CircuitState.OPEN


def test_stale_completion_does_not_pollute_new_generation():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)

    stale = breaker.try_acquire()
    for _ in range(3):
        breaker.try_acquire().record_failure()
    assert breaker.state is CircuitState.OPEN

    stale.record_success()
    assert breaker.snapshot().window_requests == 0


def test_permit_can_only_complete_once():
    breaker = CircuitBreaker(config())
    permit = breaker.try_acquire()

    assert permit.complete(CircuitOutcome.SUCCESS)
    assert not permit.complete(CircuitOutcome.FAILURE)
    snapshot = breaker.snapshot()
    assert snapshot.window_requests == 1
    assert snapshot.window_failures == 0


def test_concurrent_half_open_acquisition_respects_probe_limit():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)
    open_breaker(breaker)
    clock.advance(5.0)

    barrier = threading.Barrier(10)
    permits = []
    rejected = []
    lock = threading.Lock()

    def worker():
        barrier.wait()
        try:
            permit = breaker.try_acquire()
        except CircuitOpenError:
            with lock:
                rejected.append(True)
        else:
            with lock:
                permits.append(permit)

    threads = [threading.Thread(target=worker) for _ in range(10)]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()

    assert len(permits) == 3
    assert len(rejected) == 7
    for permit in permits:
        permit.record_success()
    assert breaker.state is CircuitState.CLOSED


def test_state_change_listener_receives_reasons_and_exceptions_are_ignored():
    clock = ManualClock()
    changes = []

    def listener(change):
        changes.append(change)
        if change.reason is StateChangeReason.COOLDOWN_ELAPSED:
            raise RuntimeError("observer failure")

    breaker = CircuitBreaker(config(), clock=clock, on_state_change=listener)
    open_breaker(breaker)
    clock.advance(5.0)
    permits = [breaker.try_acquire() for _ in range(3)]
    for permit in permits:
        permit.record_success()

    assert [change.reason for change in changes] == [
        StateChangeReason.FAILURE_RATE,
        StateChangeReason.COOLDOWN_ELAPSED,
        StateChangeReason.HALF_OPEN_SUCCEEDED,
    ]


def test_call_classifier_ignores_business_error():
    breaker = CircuitBreaker(config(minimum_requests=1))

    class BusinessError(RuntimeError):
        pass

    with pytest.raises(BusinessError):
        breaker.call(
            lambda: (_ for _ in ()).throw(BusinessError("out of stock")),
            is_failure=lambda error: not isinstance(error, BusinessError),
        )

    assert breaker.snapshot().window_requests == 0
    assert breaker.state is CircuitState.CLOSED


def test_async_call_records_results_and_uses_async_fallback():
    clock = ManualClock()
    breaker = CircuitBreaker(config(), clock=clock)

    async def scenario():
        assert await breaker.call_async(lambda: asyncio.sleep(0, result="ok")) == "ok"
        breaker.try_acquire().record_failure()
        breaker.try_acquire().record_failure()
        assert breaker.state is CircuitState.OPEN

        called = []

        async def operation():
            called.append(True)
            return "downstream"

        async def fallback(_):
            return "cached"

        assert await breaker.call_async(operation, fallback=fallback) == "cached"
        assert not called

    asyncio.run(scenario())

File: cpp/test/test_circuit_breaker.cpp
#include <gtest/gtest.h>

#include <atomic>
#include <chrono>
#include <exception>
#include <mutex>
#include <stdexcept>
#include <thread>
#include <vector>

#include "mental1104/concurrency/circuit_breaker.h"

namespace {

class ManualClock {
public:
  typedef std::chrono::steady_clock Clock;
  ManualClock() : now_(Clock::time_point()) {}

  Clock::time_point now() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return now_;
  }

  void advance(std::chrono::milliseconds duration) {
    std::lock_guard<std::mutex> lock(mutex_);
    now_ += duration;
  }

private:
  mutable std::mutex mutex_;
  Clock::time_point now_;
};

mental1104::CircuitBreakerConfig test_config() {
  mental1104::CircuitBreakerConfig config;
  config.window = std::chrono::milliseconds(10000);
  config.minimum_requests = 3;
  config.failure_rate_threshold = 0.5;
  config.slow_call_duration = std::chrono::milliseconds(1000);
  config.slow_call_rate_threshold = 0.5;
  config.open_duration = std::chrono::milliseconds(5000);
  config.half_open_max_probes = 3;
  config.half_open_successes_to_close = 3;
  return config;
}

std::unique_ptr<mental1104::CircuitBreaker>
make_breaker(const std::shared_ptr<ManualClock> &clock,
             const mental1104::StateChangeListener &listener =
                 mental1104::StateChangeListener()) {
  return std::unique_ptr<mental1104::CircuitBreaker>(
      new mental1104::CircuitBreaker(
          test_config(), listener,
          [clock]() { return clock->now(); }));
}

void open_breaker(mental1104::CircuitBreaker &breaker) {
  for (int i = 0; i < 3; ++i) {
    breaker.try_acquire().record_failure();
  }
  ASSERT_EQ(mental1104::CircuitState::Open, breaker.state());
}

} // namespace

TEST(CircuitBreakerTest, RejectsInvalidConfiguration) {
  mental1104::CircuitBreakerConfig config = test_config();
  config.window = std::chrono::milliseconds(0);
  EXPECT_THROW(mental1104::CircuitBreaker breaker(config),
               std::invalid_argument);
  config = test_config();
  config.minimum_requests = 0;
  EXPECT_THROW(mental1104::CircuitBreaker breaker(config),
               std::invalid_argument);
  config = test_config();
  config.failure_rate_threshold = 1.1;
  EXPECT_THROW(mental1104::CircuitBreaker breaker(config),
               std::invalid_argument);
  config = test_config();
  config.half_open_successes_to_close = 4;
  EXPECT_THROW(mental1104::CircuitBreaker breaker(config),
               std::invalid_argument);
}

TEST(CircuitBreakerTest, MinimumRequestsAndFailureRateOpenCircuit) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  breaker->try_acquire().record_failure();
  breaker->try_acquire().record_success();
  EXPECT_EQ(mental1104::CircuitState::Closed, breaker->state());
  breaker->try_acquire().record_failure();
  EXPECT_EQ(mental1104::CircuitState::Open, breaker->state());
}

TEST(CircuitBreakerTest, SlowSuccessfulCallsCanOpenCircuit) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  mental1104::CircuitBreakerConfig config = test_config();
  config.minimum_requests = 2;
  config.failure_rate_threshold = 1.0;
  mental1104::CircuitBreaker breaker(config, mental1104::StateChangeListener(),
                                     [clock]() { return clock->now(); });
  breaker.try_acquire().record_success();
  mental1104::CircuitPermit permit = breaker.try_acquire();
  clock->advance(std::chrono::milliseconds(1000));
  permit.record_success();
  EXPECT_EQ(mental1104::CircuitState::Open, breaker.state());
}

TEST(CircuitBreakerTest, IgnoredErrorsAreExcludedFromWindow) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  breaker->try_acquire().record_ignored();
  breaker->try_acquire().record_failure();
  const mental1104::CircuitBreakerSnapshot snapshot = breaker->snapshot();
  EXPECT_EQ(1U, snapshot.window_requests);
  EXPECT_EQ(1U, snapshot.window_failures);
  EXPECT_EQ(mental1104::CircuitState::Closed, snapshot.state);
}

TEST(CircuitBreakerTest, SlidingWindowPrunesExpiredEvents) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  mental1104::CircuitBreakerConfig config = test_config();
  config.window = std::chrono::milliseconds(2000);
  mental1104::CircuitBreaker breaker(config, mental1104::StateChangeListener(),
                                     [clock]() { return clock->now(); });
  breaker.try_acquire().record_failure();
  clock->advance(std::chrono::milliseconds(2100));
  breaker.try_acquire().record_success();
  const mental1104::CircuitBreakerSnapshot snapshot = breaker.snapshot();
  EXPECT_EQ(1U, snapshot.window_requests);
  EXPECT_EQ(0U, snapshot.window_failures);
}

TEST(CircuitBreakerTest, OpenFastFailsAndFallbackDoesNotCallDownstream) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  open_breaker(*breaker);
  std::atomic<int> calls(0);
  EXPECT_THROW(mental1104::execute(*breaker, [&calls]() {
                 ++calls;
                 return 1;
               }),
               mental1104::CircuitOpenError);
  const int value = mental1104::execute_or_fallback(
      *breaker,
      [&calls]() {
        ++calls;
        return 1;
      },
      [](const mental1104::CircuitOpenError &) { return 7; });
  EXPECT_EQ(7, value);
  EXPECT_EQ(0, calls.load());
}

TEST(CircuitBreakerTest, HalfOpenLimitsProbeRoundAndCloses) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  open_breaker(*breaker);
  clock->advance(std::chrono::milliseconds(5000));

  std::vector<mental1104::CircuitPermit> permits;
  for (int i = 0; i < 3; ++i) {
    permits.push_back(breaker->try_acquire());
  }
  EXPECT_THROW(breaker->try_acquire(), mental1104::CircuitOpenError);
  permits[0].record_success();
  permits[1].record_ignored();
  permits[2].record_success();
  EXPECT_EQ(mental1104::CircuitState::Closed, breaker->state());
}

TEST(CircuitBreakerTest, FailedOrSlowHalfOpenProbeReopens) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  open_breaker(*breaker);
  clock->advance(std::chrono::milliseconds(5000));
  breaker->try_acquire().record_failure();
  EXPECT_EQ(mental1104::CircuitState::Open, breaker->state());

  clock->advance(std::chrono::milliseconds(5000));
  mental1104::CircuitPermit permit = breaker->try_acquire();
  clock->advance(std::chrono::milliseconds(1000));
  permit.record_success();
  EXPECT_EQ(mental1104::CircuitState::Open, breaker->state());
}

TEST(CircuitBreakerTest, DuplicateAndStaleCompletionsAreIgnored) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  mental1104::CircuitPermit stale = breaker->try_acquire();
  open_breaker(*breaker);
  EXPECT_TRUE(stale.record_success());
  EXPECT_FALSE(stale.record_failure());
  EXPECT_EQ(0U, breaker->snapshot().window_requests);
}

TEST(CircuitBreakerTest, ConcurrentHalfOpenAcquisitionRespectsLimit) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::unique_ptr<mental1104::CircuitBreaker> breaker = make_breaker(clock);
  open_breaker(*breaker);
  clock->advance(std::chrono::milliseconds(5000));

  std::atomic<int> accepted(0);
  std::atomic<int> rejected(0);
  std::mutex permits_mutex;
  std::vector<mental1104::CircuitPermit> permits;
  std::vector<std::thread> threads;
  for (int i = 0; i < 10; ++i) {
    threads.push_back(std::thread([&]() {
      try {
        mental1104::CircuitPermit permit = breaker->try_acquire();
        {
          std::lock_guard<std::mutex> lock(permits_mutex);
          permits.push_back(std::move(permit));
        }
        ++accepted;
      } catch (const mental1104::CircuitOpenError &) {
        ++rejected;
      }
    }));
  }
  for (std::size_t i = 0; i < threads.size(); ++i) {
    threads[i].join();
  }
  EXPECT_EQ(3, accepted.load());
  EXPECT_EQ(7, rejected.load());
  for (std::size_t i = 0; i < permits.size(); ++i) {
    permits[i].record_success();
  }
  EXPECT_EQ(mental1104::CircuitState::Closed, breaker->state());
}

TEST(CircuitBreakerTest, ListenerFailuresDoNotBreakStateMachine) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  std::vector<mental1104::StateChangeReason> reasons;
  std::mutex reasons_mutex;
  mental1104::StateChangeListener listener = [&](const mental1104::StateChange &change) {
    {
      std::lock_guard<std::mutex> lock(reasons_mutex);
      reasons.push_back(change.reason);
    }
    if (change.reason == mental1104::StateChangeReason::CooldownElapsed) {
      throw std::runtime_error("observer failed");
    }
  };
  std::unique_ptr<mental1104::CircuitBreaker> breaker =
      make_breaker(clock, listener);
  open_breaker(*breaker);
  clock->advance(std::chrono::milliseconds(5000));
  std::vector<mental1104::CircuitPermit> permits;
  for (int i = 0; i < 3; ++i) {
    permits.push_back(breaker->try_acquire());
  }
  for (std::size_t i = 0; i < permits.size(); ++i) {
    permits[i].record_success();
  }
  EXPECT_EQ(mental1104::CircuitState::Closed, breaker->state());
  EXPECT_EQ(3U, reasons.size());
}

TEST(CircuitBreakerTest, ClassifierCanIgnoreBusinessErrorsAndVoidWorks) {
  std::shared_ptr<ManualClock> clock(new ManualClock());
  mental1104::CircuitBreakerConfig config = test_config();
  config.minimum_requests = 1;
  mental1104::CircuitBreaker breaker(config, mental1104::StateChangeListener(),
                                     [clock]() { return clock->now(); });
  mental1104::ExceptionClassifier classifier =
      [](const std::exception_ptr &) { return false; };
  EXPECT_THROW(mental1104::execute(
                   breaker,
                   []() -> int { throw std::runtime_error("out of stock"); },
                   classifier),
               std::runtime_error);
  EXPECT_EQ(0U, breaker.snapshot().window_requests);
  EXPECT_NO_THROW(mental1104::execute(breaker, []() {}));
}
